explainers/signed-exchange-subresource-subtitution-explainer.md

```diff
@@ -0,0 +1,315 @@
+1. The HTTP response of the article from the publisher's server (https://publisher.example/article.html) has an allowed-alt-sxg link header to declare
+that the UA can load the script from the prefetched signed exchange.
+   ```
+   Link: <https://cdn.publisher.example/lib.js>;
+         rel="allowed-alt-sxg";
+         header-integrity="sha256-..."
+   ```
+   So, the UA can load the JS file from the prefetched signed exchange. If the
+   publisher is providing the article in signed exchange format
+   (article.html.sxg), the allowed-alt-sxg link header is in the inner response
+   headers of the signed exchange.
+
+### Reading article while offline
+If the publisher is providing the main resource and necessary subresources in
+signed exchange format, the user of the news feed site can read the article even
+while offline. This can be done using Service Worker.
+1. The Service Worker of the news feed site fetches the main resource signed
+exchange and subresources signed exchanges (maybe using
+[Background Fetch](https://wicg.github.io/background-fetch/)) while online, and
+stores them to the CacheStorage of the feed site origin.
+1. While the user is browsing the feed site, the site inserts prefetch link
+elements to prefetch the signed exchanges. The service worker responds to the
+fetch events of the prefetch with the signed exchanges in the CacheStorage.
+1. When the user clicks a link to the article’s signed exchange, the content and
+all necessary subresources are loaded from the cached signed exchange. So the
+user can read the article while offline.
+
+
+# Proposal
+
+1. Introduce SignedExchangeCache and PrefetchedSignedExchangeCache which are
+attached to a Document. SignedExchangeCache keeps the signed exchanges which are
+prefetched from the Document. PrefetchedSignedExchangeCache keeps the signed
+exchanges which were passed from the referrer Document which triggered the
+navigation to the current Document.
+1. While processing
+"[prefetch](https://html.spec.whatwg.org/multipage/links.html#link-type-prefetch)"
+link (eg: `<link rel="prefetch"
+href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`):
+   - If succeeded prefetching the signed exchange (including merkle integrity
+   check of the body), stores the parsedExchange which is the result of
+   [parsing a signed exchange](https://wicg.github.io/webpackage/loading.html#ref-for-parsing-a-signed-exchange)
+   (= inner request URL and inner response) to the SignedExchangeCache.
+1. While
+[navigating across documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents),
+copy the parsedExchanges in the SignedExchangeCache of the source document to
+the PrefetchedSignedExchangeCache of the target document. This is intended to
+provide a way to pass the cached signed exchange across origins even if the UA
+is using origin isolated HTTPCache mechanism.
+1. While processing
```

Comment:
This seems very similar to other explainer's algorithm. Maybe add anything that's missing there and remove this one?

Reply:
Merged